Support customer-supplied encryption keys for storage

jskeet opened the pull request:
Draft of the changes to support customer-suppied encryption keys.
I haven't performed *any* testing yet, and this relies a change in the REST support libraries

evildour commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClientImpl.cs:
```diff
@@ -67,13 +69,18 @@ public static string ApplicationName
+        public StorageClientImpl(StorageService service, EncryptionKey encryptionKey = null)
         {            
             Service = GaxPreconditions.CheckNotNull(service, nameof(service));
+            EncryptionKey = encryptionKey;
```
What do you think about having this fallback to `EncryptionKey.None` so the `EncryptionKey` property is always non-null?

jskeet replied:
I wondered about that. Yes, that's probably worth doing.

jskeet commented:
The removals of ModifyRequest are just bugs left over from a previous attempt - will fix :)

evildour commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/EncryptionKey.cs:
```diff
@@ -0,0 +1,112 @@
+            }
+        }
+
+        internal void ModifyRequest(HttpRequestMessage request)
```
It might be helpful to expose this for use with the signed URL uploading: for the `ModifySessionInitiationRequest` of the options passed to `SignedUrlResumableUpload.Create`, one can just create or reuse an `EncryptionKey` instance and delegate off to its ability to modify the request.

jskeet replied:
Yup, that sounds fine to me.

jskeet commented:
Ready for actual review now - @evildour if you could have a second look to check the changes I made in the URL signer tests, that would be handy too.

evildour commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/UrlSignerTest.cs:
```diff
@@ -179,41 +179,24 @@ private void GetWithCustomerSuppliedEncryptionKeysTest_InitDelayTest()
             _fixture.RegisterDelayTest(_duration,
                 beforeDelay: async duration =>
                 {
-                    // TODO: Replace this with calls to the StorageClient if/when it supports customer-supplied encryption keys.
```
Actually, the intent was to replace everything from here down to `Assert.True(response.IsSuccessStatusCode);` with a call up UploadObjectAsync. This test isn't meant to test uploading with a signed url, it was just the easiest way to get it up there so I could test downloading (although now that I think of it, I should probably add a test for this as well: non-resumable uploading with encryption keys. I'll add this once this PR is merged).

jskeet replied:
Right - I wondered, but wasn't sure. Will see how simple that turns out to be :)

jskeet replied:
Hmm... I need to go out in 5 minutes and I can't quite see how to do this at the moment. I wonder whether we should have an option to make UrlSigner accept an EncryptionKey. What do you think?

evildour replied:
Adding the EncryptionKey as a parameter to the Sign methods on UrlSigner seems like a good idea. I can add that in a separate PR.

As for my comment about using UploadObjectAsync, I simply mean that everything from here down to line 204 should be replaced by this:
```csharp
await _fixture.Client.UpdateObjectAsync(bucket, name, new UploadObjectOptions { EncryptionKey = key });
```

jskeet replied:
Right, I'm with you. Done.

evildour commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Tests/EncryptionKeyTest.cs:
```diff
@@ -0,0 +1,113 @@
+            IEnumerable<string> values;
+            Assert.False(request.Headers.TryGetValues(EncryptionKey.AlgorithmHeader, out values));
+            Assert.False(request.Headers.TryGetValues(EncryptionKey.KeyHeader, out values));
+            Assert.False(request.Headers.TryGetValues(EncryptionKey.KeyHashHeader, out values));
```
It might be good to verify all headers are missing, to protect against copy/paste errors in future maintenance.

jskeet replied:
Do you mean the CopySource headers as well?

evildour replied:
Yes

jskeet replied:
Great, done. (All tests check all 6 headers now.)

evildour commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Tests/EncryptionKeyTest.cs:
```diff
@@ -0,0 +1,113 @@
+            IEnumerable<string> values;
+            Assert.False(request.Headers.TryGetValues(EncryptionKey.CopySourceAlgorithmHeader, out values));
+            Assert.False(request.Headers.TryGetValues(EncryptionKey.CopySourceKeyHeader, out values));
+            Assert.False(request.Headers.TryGetValues(EncryptionKey.CopySourceKeyHashHeader, out values));
```
Here too

evildour commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/CopyObjectOptions.cs:
```diff
@@ -98,6 +98,18 @@ public sealed class CopyObjectOptions
+        /// The encryption key to use for the source of the copy. The <see cref="EncryptionKey"/> and
+        /// <see cref="StorageClient.EncryptionKey"/> properties are used for the target of the copy.
+        /// </summary>
+        public EncryptionKey SourceEncryptionKey { get; set; }
```
Should this default to using the client's encryption key as well? It seems like a copy should be analogous to a download then upload using the client, but a download then upload wouldn't require specifying a key if the object were already encrypted with the client's key whereas a copy would.

jskeet replied:
Hmm. Really unsure, to be honest. It could easily be a copy from somewhere else, at which point things would break if we didn't have anything. Given that it's easy enough to specify none, I'll make that change...

evildour commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/EncryptionKey.cs:
```diff
@@ -0,0 +1,121 @@
+    /// An AES-256 key passed to the Google Cloud Storage servers via headers to allow objects to be encrypted at
+    /// rest using a client-supplied key rather than a server-supplied default key.
+    /// </summary>
+    public sealed class EncryptionKey
```
This seems like it should provide `Equals` and `GetHashCode` implementations.

jskeet replied:
I guess. It's certainly easy enough to do.

evildour commented on apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/StorageClient.cs:
```diff
@@ -54,18 +54,26 @@ public abstract partial class StorageClient
         /// The credentials are scoped as necessary.
         /// </remarks>
         /// <param name="credential">Optional <see cref="GoogleCredential"/>.</param>
+        /// <param name="encryptionKey">Optional <see cref="EncryptionKey"/> to use for all object-based operations by default.</param>
```
Nit: may want to add "May be null" to match the async version.

jskeet replied:
Done.

jskeet commented:
I think I've addressed all the issues other than the UrlSignerTest, which I'll try to look at tomorrow. (Or we could merge as-is, maybe with a TODO...)

jskeet commented:
Thanks for the review and feedback. All done and squashed into one commit. Will merge on green.
